Remove account lockout from failed recovery attempts

jentfoo opened the pull request:
This account lockout looks to be unnecessary and potentially problematic.  Recovery codes and recovery through MFA are not possible to brute force.

In addition the potential to lockout an account from being able to use a recovery method could result in them being unable to unlock their account from other potential abuse cases (for example an attacker locking the account from failed password attempts).

For that reason I believe we should remove this lock as part of Teleport 15.  Opening this PR for discussion, please highlight if there are any concerns which motivated this account lockout and which I may have missed.

codingllama commented on lib/auth/accountrecovery_test.go:
```diff
@@ -402,10 +374,6 @@ func TestVerifyAccountRecovery_WithAuthnErrors(t *testing.T) {
 
-			attempts, err := srv.Auth().GetUserRecoveryAttempts(ctx, u.username)
-			require.NoError(t, err)
-			require.Len(t, attempts, 1)
```
Should we keep a modified version of this assertion to ensure that attempts are still recorded, or that covered elsewhere?

jentfoo replied:
With this change there was no need to retrieve the attempts.  So eventually I expected this API will be removed from the Auth.  I considered trying to add in an API just for test, but I don't feel this check adds much value.  The general account behaviors are already covered.

codingllama commented on lib/auth/accountrecovery.go:
```diff
@@ -117,13 +103,6 @@ func (a *Server) verifyCodeWithRecoveryLock(ctx context.Context, username string
 	}
 
-	status := user.GetStatus()
-	if status.IsLocked && status.RecoveryAttemptLockExpires.After(a.clock.Now().UTC()) {
```
Can we delete the Login.RecoveryAttemptLockExpires field?

jentfoo replied:
We can, however I was trying to avoid protocol changes until the next major version.  Can I simply remove this field now?

codingllama replied:
It depends on how the field is usage. We can deprecate it and remove in the following major, but sometimes outright removal is possible.

codingllama commented on lib/services/identity.go:
```diff
@@ -267,12 +267,6 @@ type Identity interface {
 	CreateUserRecoveryAttempt(ctx context.Context, user string, attempt *types.RecoveryAttempt) error
 
-	// GetUserRecoveryAttempts returns user recovery attempts sorted by oldest to latest time.
-	GetUserRecoveryAttempts(ctx context.Context, user string) ([]*types.RecoveryAttempt, error)
```
How does one read recovery attempts, as a Teleport user/admin?

Are there leftover methods, meant to implement this interface, that we could now remove?

jentfoo replied:
You bring up a good point, maybe I should bring this API back?

codingllama replied:
I thought we wanted the attempts tracked somewhere, but maybe they are already in audit? I don't recall from the top of my head how this plugs into the wider system.

jentfoo replied:
We do log failures here: https://github.com/gravitational/teleport/pull/35325/files#diff-58ee6a3f9df8cbca83069a64e445a0b91c16352c77a92f8cdec8be15847c7d89L362

However I am not sure if we have another need to lookup current failures on an account?

codingllama commented on rfd/0029-account-lifecycle.md:
```diff
@@ -56,8 +56,6 @@ domain names in the prefix about leaked tokens.
 * [ ] Each token can only be used one time successfully or unsuccessfully.
 * [ ] Teleport should store tokens hashed with `crypto/bcrypt` and compare using constant-time compare.
 Teleport should store the time when the tokens were generated alongside the hashed tokens.
-* [ ] Trying a wrong token 3 times in a row should lead to a temporary account lock following
```
@klizhentas or @kimlisa, do you remember the reasoning behind this requirement?

Did we have brute-force concerns with some of the recovery flows? Say, leaked token + guessed password to register a new MFA device, or something like that?

russjones reviewed (dismissed):
@jentfoo Can you link me to the RFD where this change was discussed and approved?

russjones commented:
@jentfoo Is this tied to a compliance goal or issue that will reach it's SLA if not addressed?

russjones commented:
@jentfoo Please propose changes via our RFD process if you want to change the behavior of Teleport.

jentfoo commented:
@russjones I submitted the RFD under the PR here: https://github.com/gravitational/teleport/pull/35533

The RFD is fairly light in content because this is a relatively isolated change, but hopefully it will provides a better description of why we want to make this improvement.  Please let me know your thoughts and questions, thank you!

jentfoo commented:
@zmb3 and @codingllama, I have partially updated this PR.  This will depend on the `e` update I just submitted here: https://github.com/gravitational/teleport.e/pull/3118

I would appreciate a look around the API changes (mostly captured in the most recent commit).  I believe I have removed all possible API as discussed [in the RFD](https://github.com/gravitational/teleport/pull/35533).  I have kept the `RecoveryAttemptLockExpires` field in the protobuf but renamed as `DEPRECATED`, let me know if there is a more standard way to maintain compatibility with these changes.  Thank you!

zmb3 commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -3265,10 +3265,9 @@ message LoginStatus {
-  // from reaching MaxAccountRecoveryAttempts. This field is used to determine
-  // if a user got locked from recovery attempts.
-  google.protobuf.Timestamp RecoveryAttemptLockExpires = 5 [
+  // DEPRECATED RecoveryAttemptLockExpires contained the time when this lock will expire
```
You're going to want to use reserved fields for this: https://protobuf.dev/programming-guides/proto3/#fieldreserved

codingllama replied:
Adding to what Zac said, but without looking at the PR in its entirety:

* If you still need to keep the field around, use the [deprecated=true] field option. Some folks add a "// Deprecated: \<reason\>" comment at the end of the field docs, in Go style, but that's optional.

* If you don't need to access the field, you can reserve both its tag (number) and name, effectively removing it from the proto.

jentfoo replied:
Thank you for the advice, since we don't need to reference the field I have updated it to be `reserved`

strideynet commented on integrations/operator/crdgen/testdata/protofiles/teleport/legacy/types/types.proto:
```diff
@@ -3081,14 +3081,8 @@ message LoginStatus {
-    (gogoproto.nullable) = false,
-    (gogoproto.jsontag) = "recovery_attempt_lock_expires,omitempty"
-  ];
+  reserved 5;  // removed RecoveryAttemptLockExpires after lockout was removed
```
Usually I see these reserve blocks at the top of the message - but unsure if that's a standard thing.

codingllama replied:
AFAIK this goes either way. Not sure if we have a consensus.

jentfoo commented:
With the `e` update, this should be ready for review now.  I don't plan to backport this to v14 or older, but I would like this change to be included in v15.

I believe the flakey test error is just due to the size / timeout.  I have tried to look for any test failures and I can't find any.

Thank you for all the feedback so far!

Tener commented on lib/auth/accountrecovery.go:
```diff
@@ -106,7 +92,7 @@ func (a *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccou
 // locked from logging in. Modeled after existing function WithUserLock.
 func (a *Server) verifyCodeWithRecoveryLock(ctx context.Context, username string, recoveryCode []byte) error {
-	user, err := a.Services.GetUser(ctx, username, false)
+	_, err := a.Services.GetUser(ctx, username, false)
```
I think that the method description needs updating?

jentfoo replied:
Thank you, updated!

Tener reviewed (approved):
Flaky test fails due to: `panic: test timed out after 10m0s`; it may be hard to make the test faster, in which case an override will be needed.

codingllama commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -3265,14 +3265,8 @@ message LoginStatus {
-    (gogoproto.nullable) = false,
-    (gogoproto.jsontag) = "recovery_attempt_lock_expires,omitempty"
-  ];
+  reserved 5; // removed RecoveryAttemptLockExpires after lockout was removed
```
Suggestion: I typically state the type too. Feel free to disregard.

```suggestion
  reserved 5; // removed "google.protobuf.Timestamp RecoveryAttemptLockExpires" after lockout was removed
```

codingllama commented on integrations/operator/crdgen/testdata/protofiles/teleport/legacy/types/types.proto:
```diff
@@ -3081,14 +3081,8 @@ message LoginStatus {
-    (gogoproto.nullable) = false,
-    (gogoproto.jsontag) = "recovery_attempt_lock_expires,omitempty"
-  ];
+  reserved 5;  // removed RecoveryAttemptLockExpires after lockout was removed
```
Same suggestion as above:

```suggestion
  reserved 5;  // removed "google.protobuf.Timestamp RecoveryAttemptLockExpires" after lockout was removed
```

codingllama commented on lib/auth/accountrecovery.go:
```diff
@@ -132,24 +110,10 @@ func (a *Server) verifyCodeWithRecoveryLock(ctx context.Context, username string
-	switch {
-	case err != nil:
+	if err := a.recordFailedRecoveryAttempt(ctx, username); err != nil {
 		log.Error(trace.DebugReport(err))
```
Should we add some context to this message, or does the underlying error carry everything we need?

Eg:

```suggestion
		log.WithError(err).Warn("Errored recording an account recovery failure")
```

codingllama commented on lib/auth/accountrecovery.go:
```diff
@@ -307,79 +271,28 @@ func (a *Server) verifyAuthnWithRecoveryLock(ctx context.Context, startToken typ
-	}
-
-	// Delete all tokens related to this user, to force user to restart the recovery flow.
-	if err := a.deleteUserTokens(ctx, startToken.GetUser()); err != nil {
```
Should we keep this part? Does it have to do with recovery attempts too?

jentfoo replied:
Thank you for mentioning this.  I would like to hear @kimlisa's perspective on this removal.

This condition made no sense to me, having a reset failure should not allow any impacts to the account itself.  These tokens are used to generate new recovery codes, so potential impact cases are unlikely, but the behavior of allowing a potentially unauthorized actor to impact the account seems wrong to me.

I also made a recent relevant change to ensure these tokens are always deleted after use: https://github.com/gravitational/teleport/pull/36527

So in theory these tokens should be very short lived (unless the client was interrupted or otherwise stopped the token generation process for some reason)

I believe because of that, this was in error, but maybe @kimlisa is aware of something I am missing.

jentfoo replied:
The tokens being deleted in this line are tokens used to generate new recovery codes, not the recovery code itself.

> the particular token in use here will always be deleted

The recovery code is removed after use here: https://github.com/gravitational/teleport/blob/f356242a2c30ca28f6326d175adeb8ac22655706/lib/auth/accountrecovery.go#L146-L151

The tokens used for generating recovery codes will be deleted after successful use (as of the linked PR above), but I am not aware of a deletion process if they go unused (maybe there is a TTL I have not seen)

> but we won't delete any other recovery tokens to avoid a situation where a malicious user keeps making someone else recovery attempts fail through this side-effect?

This is correct as of this change.  But to ensure the potential impact is highlighted, it would only impact a user attempting to regenerate their recovery codes while an attacker was doing an invalid recovery attempt.  It is a very small risk, but I still believe this change is more technically correct.

@kimlisa Let me know if I have any misunderstandings, thank you!

kimlisa replied:
i didn't really understand what error the prior code had 😅 (i would very much like to understand though).

the start token used here expired in 3 hrs (it's the token that's part of the recovery URL we send to users email after they provided us with a valid recovery code)

in the prior code, the `delete token` block was only ran if the user reached max attempts at providing us their password/2ndfactor. i thought it appropriate then to [delete all user tokens](https://github.com/gravitational/teleport/blob/5f9803917c3daecb9caec434612be0e4b5f31231/lib/auth/usertoken.go#L44) so they can't `attempt anymore with this token` and the user is going to get locked anyways.

if the user tried to access the URL with the deleted token, it'll show something like "expired or invalid token", prompting the user to try another recovery code. 

i think if we didn't delete it, the way it was coded, it will allow the user to keep attempting?

but since we are removing max attempts and locks we no longer need the `delete tokens`, we just allow the user to keep trying it looks like?

jentfoo replied:
Thank you @kimlisa, I did not realize this was also deleting the token used in the email flow.  That said I believe this should still be removed, consider the following:

1. A victim is under targeted attack, their account is locked and they can not sign in.  They initiate a recovery process to unlock their account and attempt to gain immediate access.
2. The attacker is persistent and is simultaneously and continually doing false recovery attempts.  Resulting in the deletion of the initiated recovery email token.
3. The victim has no ability to recover their account unless they can get lucky enough to complete the entire process before the attacker can invalidate new tokens (or they end up using all their recovery codes trying).

Any token invalidation from an unauthenticated and untrusted party has some potential to lead to a denial of service.  For that reason this seems potentially risky with or without the recovery lockout removal.

> i think if we didn't delete it, the way it was coded, it will allow the user to keep attempting?

Maybe, this is not clear to me.  I expected to see this action as part of https://github.com/gravitational/teleport/blob/master/lib/auth/accountrecovery.go#L386, however I am not sure where this token is being deleted after use?  I just verified it's being deleted somewhere, and I don't see any way this line was the delete responsible to prevent reuse of the email token.

Let me know your thoughts and questions, thank you very much @kimlisa!

kimlisa replied:
got it, thanks for explaining!

> however I am not sure where this token is being deleted after use

if you've successfully authenticated with the start token, then we delete it afterwards [here](https://github.com/gravitational/teleport/blob/a15876724067837ba0449f06f01b9b0a25c370d0/lib/auth/accountrecovery.go#L283). i think it makes sense to delete it right afterwards in case user for some reason doesn't make it to completing the recovery (since it's a longer living token)

if you failed, you will get unlimited attempts (just confirming that's how it's currently going to work now)

We also delete all user tokens when we finish [creating new recovery codes](https://github.com/gravitational/teleport/blob/a15876724067837ba0449f06f01b9b0a25c370d0/lib/auth/accountrecovery.go#L491) (which is after `completing account recovery` (this portion just removes locks and allows you to reset password or second factor))

jentfoo replied:
Thank you, that makes sense!

codingllama commented on lib/auth/accountrecovery.go:
```diff
@@ -307,79 +271,28 @@ func (a *Server) verifyAuthnWithRecoveryLock(ctx context.Context, startToken typ
-	// Restart the attempt counter, to not block users from trying again with another recovery code.
-	if err := a.DeleteUserRecoveryAttempts(ctx, startToken.GetUser()); err != nil {
+	if err := a.recordFailedRecoveryAttempt(ctx, startToken.GetUser()); err != nil {
 		log.Error(trace.DebugReport(err))
```
Ditto for log message.

jentfoo replied:
Done, thank you!

codingllama commented on lib/auth/accountrecovery.go:
```diff
@@ -307,79 +271,28 @@ func (a *Server) verifyAuthnWithRecoveryLock(ctx context.Context, startToken typ
 	// Record and log failed attempt.
 	now := a.clock.Now().UTC()
 	attempt := &types.RecoveryAttempt{Time: now, Expires: now.Add(defaults.AttemptTTL)}
 	if err := a.CreateUserRecoveryAttempt(ctx, username, attempt); err != nil {
```
Maybe just:

```suggestion
	return trace.Wrap(a.CreateUserRecoveryAttempt(ctx, username, attempt))
```

jentfoo replied:
Done, thank you!

codingllama reviewed (commented):
Seems reasonable to me, but I would ask @kimlisa to take a look just in case.

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -102,11 +88,10 @@ func (a *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccou
+// the code is valid.  If the code is invalid, a failed recovery attempt will be recorded.
 func (a *Server) verifyCodeWithRecoveryLock(ctx context.Context, username string, recoveryCode []byte) error {
-	user, err := a.Services.GetUser(ctx, username, false)
+	_, err := a.Services.GetUser(ctx, username, false)
```
the method name `verifyCodeWithRecoveryLock` is no longer correct since we don't lock anymore. maybe... `verifyCodeWithRecord`?

jentfoo replied:
Updated, thank you!

kimlisa commented on lib/auth/accountrecovery.go:
```diff
@@ -288,13 +252,12 @@ func (a *Server) VerifyAccountRecovery(ctx context.Context, req *proto.VerifyAcc
-// and all user's tokens are deleted. Modeled after existing function WithUserLock.
+// verifyAuthnWithRecoveryLock validates the recovery code (through authenticateFn).  If the code is invalid,
+// a failed recovery attempt will be recorded.
 func (a *Server) verifyAuthnWithRecoveryLock(ctx context.Context, startToken types.UserToken, authenticateFn func() error) error {
```
same here to (the method name)

jentfoo replied:
Updated, thank you!

jentfoo committed:
accountrecovery: Update `WithLock` function names

jentfoo committed:
Remove account lockout from failed recovery attempts

This account lockout looks to be unecessary and potentially problematic.  Recovery codes and recovery through MFA are not possible to brute force.

In addition the potential to lockout an account from being able to use a recovery method could result in them being unable to unlock their account from other potential abuse cases (for example an attacker locking the account from failed password attempts).

As discussed in the RFD (https://github.com/gravitational/teleport/pull/35533) this includes the removal of all the API used for this locking mechanism.

jentfoo committed:
accountrecovery: Combine verifyRecoveryCode and verifyRecoveryCodeWithRecord into one function

jentfoo commented on lib/auth/accountrecovery.go:
```diff
@@ -102,11 +88,10 @@ func (a *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccou
-	user, err := a.Services.GetUser(ctx, username, false)
+// verifyRecoveryCodeWithRecord validates the recovery code for the user and will unlock their account if
+// the code is valid.  If the code is invalid, a failed recovery attempt will be recorded.
+func (a *Server) verifyRecoveryCodeWithRecord(ctx context.Context, username string, recoveryCode []byte) error {
```
I have rebased on the latest, removing the `e` update (already updated) and merged the `verifyRecoveryCode` and `verifyRecoveryCodeWithRecord` into a single function.

The auditing is now entirely colocated within a `defer func` which checks the status.  I maintained the behavior exactly as it is today, but as mentioned on DM I would appreciate a look to confirm auditing looks good.

Thank you for the help!

zmb3 commented on lib/auth/accountrecovery.go:
```diff
@@ -103,66 +89,61 @@ func (a *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccou
-	case err != nil:
+// verifyRecoveryCodeWithRecord validates the recovery code for the user and will unlock their account if
+// the code is valid.  If the code is invalid, a failed recovery attempt will be recorded.
+func (a *Server) verifyRecoveryCodeWithRecord(ctx context.Context, username string, recoveryCode []byte) (errResult error) {
```
```suggestion
func (a *Server) verifyRecoveryCode(ctx context.Context, username string, recoveryCode []byte) (errResult error) {
```

Now that you don't have two methods that do the same thing, we can simplify the name here.

jentfoo replied:
Updated, thank you

zmb3 commented on lib/auth/accountrecovery.go:
```diff
@@ -103,66 +89,61 @@ func (a *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccou
+			}
+		} else {
+			event.Metadata.Code = events.RecoveryCodeUseFailureCode
+			traceErr := trace.NotFound("invalid user or user does not have recovery codes")
+
+			if hasRecoveryCodes {
+				traceErr = trace.BadParameter("recovery code did not match")
+			}
+
+			event.Status.Error = traceErr.Error()
+			event.Status.UserMessage = traceErr.Error()
```
Do we need to create an error object only to pull the string out of it?

Creating an error with the trace package is a costly operation since it pulls a full stack trace. If there are recovery codes, we're collecting not one but two stack traces here, only to throw them away.

I would get rid of `traceErr` completely, and just set `event.Status.Error = "recovery code did not match"`.

jentfoo replied:
I just moved the code without change.  In the most recent commit I updated this to only set the `event.Status.Error` string directly.

zmb3 commented on lib/auth/accountrecovery.go:
```diff
@@ -308,79 +258,24 @@ func (a *Server) verifyAuthnWithRecoveryLock(ctx context.Context, startToken typ
-	maxedAttempts := true
-
+// recordFailedRecoveryAttempt creates and inserts a recovery attempt and if user has reached max failed attempts.
+func (a *Server) recordFailedRecoveryAttempt(ctx context.Context, username string) error {
```
Do we even need this anymore? What is the point of recording a failed audit attempt if we no longer need to know how many failed attempts there have been?

The failed attempt is already written to the audit log, so I suspect this can be removed.

jentfoo replied:
I removed the attempt CRUD API fully now, thank you for the help here

zmb3 commented on lib/services/identity.go:
```diff
@@ -269,12 +269,6 @@ type Identity interface {
 	CreateUserRecoveryAttempt(ctx context.Context, user string, attempt *types.RecoveryAttempt) error
 
-	// GetUserRecoveryAttempts returns user recovery attempts sorted by oldest to latest time.
-	GetUserRecoveryAttempts(ctx context.Context, user string) ([]*types.RecoveryAttempt, error)
```
Can we get rid of `GetUserRecoveryAttempts` completely? I see you removed it from the interface, but the implementation remains, so we have an orphaned function that nothing is calling left behind.

jentfoo replied:
This was an oversight, removed, thank you!

zmb3 replied:
I still see some references to `GetUserRecoveryAttempts`, `DeleteUserRecoveryAttempts`, and `CreateUserRecoveryAttempt` in the codebase. These can be removed, along with the `recoveryAttemptsPrefix` constant.

jentfoo replied:
I removed `recoveryAttemptsPrefix`, that was missed.  But can you point me to the other cases?  I just did a grep for the three CRUD functions and can't find any reference to any of them.  Thank you!

jentfoo committed:
Further GRPC cleanup after PR feedback

zmb3 commented on lib/auth/accountrecovery.go:
```diff
@@ -103,66 +88,54 @@ func (a *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccou
-		return a.verifyRecoveryCode(ctx, username, recoveryCode)
-	case err != nil:
+// verifyRecoveryCode validates the recovery code for the user and will unlock their account if
+// the code is valid.  If the code is invalid, a failed recovery attempt will be recorded.
```
```suggestion
// the code is valid.
```

We're no longer recording failed attempts (at least not in the way we used to).

jentfoo replied:
We still are recording an audit event, but no strong preference from me, removed.

zmb3 commented on lib/auth/accountrecovery.go:
```diff
@@ -289,13 +232,12 @@ func (a *Server) VerifyAccountRecovery(ctx context.Context, req *proto.VerifyAcc
-func (a *Server) verifyAuthnWithRecoveryLock(ctx context.Context, startToken types.UserToken, authenticateFn func() error) error {
+// verifyAuthnRecoveryWithRecord validates the recovery code (through authenticateFn).  If the code is invalid,
+// a failed recovery attempt will be recorded.
+func (a *Server) verifyAuthnRecoveryWithRecord(ctx context.Context, startToken types.UserToken, authenticateFn func() error) error {
```
```suggestion
func (a *Server) verifyAuthnRecovery(ctx context.Context, startToken types.UserToken, authenticateFn func() error) error {
```

Similar to the previous change, we can get rid of the `WithRecord` suffix.

jentfoo replied:
Updated, thank you

zmb3 reviewed (commented):
We're almost there.

jentfoo committed:
Apply PR Feedback

zmb3 commented:
/excludeflake *

codingllama commented on api/proto/teleport/legacy/types/types.proto:
```diff
@@ -3276,14 +3276,8 @@ message LoginStatus {
-    (gogoproto.nullable) = false,
-    (gogoproto.jsontag) = "recovery_attempt_lock_expires,omitempty"
-  ];
+  reserved 5; // removed "google.protobuf.Timestamp RecoveryAttemptLockExpires" after lockout was removed
+  reserved "RecoveryAttemptLockExpires";
```
The removal of this field is making [UpdateAndSwapUser][1] fail.

If the stored user had the "recovery_attempt_lock_expires" field set, then the comparison always fails. That is because when we read and unmarshal the user [the unknown field gets dropped][2], but the actual CompareAndSwap operation compares [using the storage blobs instead][3].

Reverting the deletion, but keeping the field deprecated should serve as an immediate fix. Arguably, a better fix is to make sure CompareAndSwap compares apples-to-apples: either user vs user, or blob vs blob, but never user vs blob.

@jentfoo @rosstimothy

[1]: https://github.com/gravitational/teleport/blob/b8deacfaffc25073f22ff0dadf157344b2fd2688/lib/services/local/users.go#L383

[2]: https://github.com/gravitational/teleport/blob/b8deacfaffc25073f22ff0dadf157344b2fd2688/lib/services/local/users.go#L384

[3]: https://github.com/gravitational/teleport/blob/b8deacfaffc25073f22ff0dadf157344b2fd2688/lib/backend/lite/lite.go#L461

jentfoo replied:
I just submitted a PR to bring the field back: https://github.com/gravitational/teleport/pull/37618

Thank you for finding this issue @codingllama!
